src/coreclr/pal/src/exception/remote-unwind.cpp

```diff
@@ -2126,6 +2153,16 @@ PAL_VirtualUnwindOutOfProc(CONTEXT *context, KNONVOLATILE_CONTEXT_POINTERS *cont
 #elif defined(TARGET_ARM64)
     TRACE("Unwind: pc %p sp %p fp %p\n", (void*)context->Pc, (void*)context->Sp, (void*)context->Fp);
     result = GetProcInfo(context->Pc, &procInfo, &info, &step, false);
+    if (result && step)
+    {
+        // If the PC is at the start of the function and the unwind encoding is frameless, back up PC by 1 to the previous
+        // function and get the unwind info for that function. This for functions that have been optimized to tail call.
+        if ((context->Pc == procInfo.start_ip) && (procInfo.format & UNWIND_ARM64_MODE_MASK) == UNWIND_ARM64_MODE_FRAMELESS)
```

Comment:
Is it always the case that PC at start of a function that is frameless is related to the previous function? What if there is a standalone frameless function that's not right at the end of another one? And cannot if the one before be actually unrelated? 
In your example, it is not clear to me how we can end up using an incorrect LR for unwinding. It seems we are at a spot in the code where we should use the LR from the current context as the unwound PC since there is nothing pushed on the stack.

Reply:
Yes, the encoding is 0x02000000 which means it is frameless and there is nothing on SP. The problem is that LR at this point is the return address from the previous frame (#2).  The unwind stopped because SP and IP didn't change. When I studied what lldb did for this stack trace, it had logic to try IP - 1. 

frame #3 is frame in question (SP FP PC).  

```
frame #0: 0x000000016d4dd470 0x000000016d4dd4f0 0x00000001a6c24930 libsystem_kernel.dylib`__wait4 + 8
frame #1: 0x000000016d4dd470 0x000000016d4dd4f0 0x000000010333d554 libcoreclr.dylib`PROCCreateCrashDumpIfEnabled + 1392
frame #2: 0x000000016d4dd500 0x000000016d4dd510 0x000000010333a8a4 libcoreclr.dylib`PROCAbort + 52
frame #3: 0x000000016d4dd520 0x000000016d4dd540 0x000000010333a7bc libcoreclr.dylib`PROCEndProcess(void*, unsigned int, int) + 396
frame #4: 0x000000016d4dd550 0x000000016d4dda80 0x0000000103565d0c libcoreclr.dylib`UnwindManagedExceptionPass1(PAL_SEHException&, _CONTEXT*) + 968
frame #5: 0x000000016d4dda90 0x000000016d4dde40 0x0000000103565d54 libcoreclr.dylib`DispatchManagedException(PAL_SEHException&, bool) + 68
```

Here is the PAL unwind logging without this hack:

```
PAL: pc 0x10333a8a4 sp 0x16d4dd500 fp 0x16d4dd510 lr 0x10333a8a4                   (frame 2 abpve)
Unwind 0x1037ba294 offset 000368a4 ver 1 common off: 0000001c common cnt: 45 pers off: 000000d0 pers cnt: 2 index off: 000000d8 index cnt: 13
Second level compressed: funcStart 0x10333a7d8 funcEnd 0x10333a8ac pageEntry 01032cd8 pageEntryNext 00032dac
Second level compressed common table: 04000001 for offset 00032da4 encodingIndex 1
Encoding 04000001 start 0x10333a7d8 end 0x10333a8ac found for ip 0x10333a8a4
Compact step 0x10333a8a4 format 04000001 start_ip 0x10333a7d8
BEFORE: addr 0x32dac0000000d lr 0x10333a8a4 fp 0x16d4dd510
AFTER: addr 0x16d4dd508 lr 0x10333a7bc fp 0x16d4dd540
SUCCESS: compact step encoding 04000001 pc 0x10333a7bc sp 0x16d4dd520 fp 0x16d4dd540 lr 0x10333a7bc
Unwind: sp 0x16d4dd500 ip 0x10333a8a4 off 000000cc mod 0x103304000

PAL: pc 0x10333a7bc sp 0x16d4dd520 fp 0x16d4dd540 lr 0x10333a7bc
Unwind 0x1037ba294 offset 000367bc ver 1 common off: 0000001c common cnt: 45 pers off: 000000d0 pers cnt: 2 index off: 000000d8 index cnt: 13
Second level compressed: funcStart 0x10333a7bc funcEnd 0x10333a7c4 pageEntry 00032cbc pageEntryNext 03032cc4
Second level compressed common table: 02000000 for offset 00032cbc encodingIndex 0
Encoding 02000000 start 0x10333a7bc end 0x10333a7c4 found for ip 0x10333a7bc
Compact step 0x10333a7bc format 02000000 start_ip 0x10333a7bc
SUCCESS: compact step encoding 02000000 pc 0x10333a7bc sp 0x16d4dd520 fp 0x16d4dd540 lr 0x10333a7bc
Unwind: sp 0x16d4dd520 ip 0x10333a7bc off 00000000 mod 0x103304000
Unwind: sp not increasing or ip == 0 sp 0x16d4dd520 ip 0x10333a7bc
```

This is another case where it isn't clear what is supposed to happen.  So I tried to emulated what LLDB does in a very limited fashion.

Reply:
I looked at this a little more and it is only if PC - 1 is a BL instruction that this logic should be executed (along with the other 2 conditions).  It looks like the only way PC = <start of TerminateProcess> is if the BL at the end of PROCEndProcess was executed.

Reply:
Now I need to know about to decode an arm64 BL opcode.  It isn't that clear on the internet.

Reply:
This site has a table of instruction encodings for arm64:
http://kitoslab-eng.blogspot.com/2012/10/armv8-aarch64-instruction-encoding.html

Reply:
I've added a check for BL opcode as the previous instruction.